src/Client/ConcordiumClient.cs

```diff
@@ -543,5 +545,94 @@ await Task.WhenAll(response.ResponseHeadersAsync, response.ResponseAsync)
             .ConfigureAwait(false);
     }
 
+    /// <summary>
+    /// Get all smart contract modules that exist at the end of a given block.
+    /// </summary>
+    /// <param name="blockHashInput">Block hash from where modules will be given.</param>
+    /// <param name="token">Cancellation token</param>
+    /// <returns>Stream of modules in given block.</returns>
+    /// <exception cref="RpcException">
+    /// RPC error occurred, access <see cref="RpcException.StatusCode"/> for more information.
+    /// <see cref="StatusCode.NotFound"/> indicates block was not found.
+    /// </exception>
+    public Task<QueryResponse<IAsyncEnumerable<ModuleReference>>> GetModuleListAsync(IBlockHashInput blockHashInput, CancellationToken token = default)
+    {
+        var response = this.Raw.GetModuleList(blockHashInput.Into(), token);
+        return QueryResponse<IAsyncEnumerable<ModuleReference>>.From(response, ModuleReference.From, token);
+    }
+
+    /// <summary>
+    /// Get the addresses of all smart contract instances in a given block.
+    /// </summary>
+    /// <param name="blockHashInput">Block hash from where contracts will be given.</param>
+    /// <param name="token">Cancellation token</param>
+    /// <returns>Stream of contracts in given block.</returns>
+    /// <exception cref="RpcException">
+    /// RPC error occurred, access <see cref="RpcException.StatusCode"/> for more information.
+    /// <see cref="StatusCode.NotFound"/> indicates block was not found.
+    /// </exception>
+    public Task<QueryResponse<IAsyncEnumerable<ContractAddress>>> GetInstanceList(IBlockHashInput blockHashInput, CancellationToken token = default)
+    {
+        var response = this.Raw.GetInstanceList(blockHashInput.Into(), token);
+        return QueryResponse<IAsyncEnumerable<ContractAddress>>.From(response, ContractAddress.From, token);
+    }
+
+    /// <summary>
+    /// Returns the source of a smart contract module.
+    /// </summary>
+    /// <param name="blockHashInput">Block hash from where module source will be given.</param>
+    /// <param name="moduleReference">Module reference from where source is requested.</param>
+    /// <param name="token">Cancellation token</param>
+    /// <returns>Mapped versioned module source instances.</returns>
+    /// <exception cref="RpcException">
+    /// RPC error occurred, access <see cref="RpcException.StatusCode"/> for more information.
+    /// <see cref="StatusCode.NotFound"/> indicates block was not found.
+    /// </exception>
+    public async Task<QueryResponse<VersionedModuleSource>> GetModuleSourceAsync(IBlockHashInput blockHashInput, ModuleReference moduleReference, CancellationToken token = default)
+    {
+        var response = this.Raw.GetModuleSourceAsync(new ModuleSourceRequest { BlockHash = blockHashInput.Into(), ModuleRef = moduleReference.Into() }, token);
+
+        await Task.WhenAll(response.ResponseHeadersAsync, response.ResponseAsync)
+            .ConfigureAwait(false);
```

Comment:
When using `Task.WhenAll` we schedule both task at the same time and they can run asynchronously. When we then later in the code call await it returns immediately since the task has already been executed.

If we called them sequentially we wouldn’t use the potential of having the task running async at the same time.

Reply:
Okay, I imagine these tasks are already started, but cannot find any documentation on this, so I guess it is fine to keep this for now.

Reply:
I was to fast on this one - they are both started. I just needed to look at the code again to remember why I did like this. 

The pattern was to make it explicit that I in the end wanted to wait on both. I didn't want to pass any mapping delegate of the result to the `QueryResponse` and since the header part was the same for all responses I wanted to avoid code duplication.

It could maybe been have done differently.

Reply:
Let's just leave it for now, but if you find the reason, please let me know, ideally it would be in a comment in the code